internal/datatransfer/monitor.go

```diff
@@ -0,0 +1,107 @@
+package datatransfer
+
+import (
+	"git.sr.ht/~spc/go-log"
+	"github.com/jakub-dzon/k4e-device-worker/internal/configuration"
+	"github.com/jakub-dzon/k4e-device-worker/internal/datatransfer/s3"
+	"github.com/jakub-dzon/k4e-device-worker/internal/workload"
+	"github.com/jakub-dzon/k4e-operator/models"
+	"path"
+	"sync"
+	"time"
+)
+
+type Monitor struct {
+	workloads                   *workload.WorkloadManager
+	config                      *configuration.Manager
+	ticker                      *time.Ticker
+	lastSuccessfulSyncTimes     map[string]time.Time
+	lastSuccessfulSyncTimesLock sync.RWMutex
+}
+
+func NewMonitor(workloadsManager *workload.WorkloadManager, configManager *configuration.Manager) *Monitor {
+	ticker := time.NewTicker(configManager.GetDataTransferInterval())
+	monitor := Monitor{
+		workloads:                   workloadsManager,
+		config:                      configManager,
+		ticker:                      ticker,
+		lastSuccessfulSyncTimes:     make(map[string]time.Time),
+		lastSuccessfulSyncTimesLock: sync.RWMutex{},
+	}
+	return &monitor
+}
+
+func (m *Monitor) Start() {
+	go func() {
+		for range m.ticker.C {
+			m.syncPaths()
+		}
+	}()
+}
+
+func (m *Monitor) GetLastSuccessfulSyncTime(workloadName string) *time.Time {
+	m.lastSuccessfulSyncTimesLock.RLock()
+	defer m.lastSuccessfulSyncTimesLock.RUnlock()
+	if t, ok := m.lastSuccessfulSyncTimes[workloadName]; ok {
+		return &t
+	}
+	return nil
+}
+
+func (m *Monitor) WorkloadRemoved(workloadName string) {
+	m.lastSuccessfulSyncTimesLock.Lock()
+	defer m.lastSuccessfulSyncTimesLock.Unlock()
+	delete(m.lastSuccessfulSyncTimes, workloadName)
+}
+func (m *Monitor) syncPaths() {
+	workloads, err := m.workloads.ListWorkloads()
+	if err != nil {
+		log.Errorf("Can't get the list of workloads: %v", err)
+	}
+	if len(workloads) == 0 {
+		return
+	}
+	storage := m.config.GetDeviceConfiguration().Storage
+	if storage != nil && storage.S3 != nil {
```

Comment:
perhaps an early return if storage is nil or storage.s3 is nil to avoid one nesting level ?